### PR TITLE
Fix VAC crew hash aggregation for multi-member groups

### PR DESCRIPTION
**Pull request by JFlo21**
## Summary
Fixed a bug where edits to VAC crew fields (name/dept/job) on non-first-sorted rows were silently ignored, causing Excel files to skip regeneration even when criteria were met. The root cause was that hash metadata was only read from the first sorted row, but VAC crew groups are not split per foreman and can contain multiple crew members.

## Key Changes

- **Hash aggregation fix**: Modified `calculate_data_hash()` to aggregate `__vac_crew_name`, `__vac_crew_dept`, and `__vac_crew_job` across ALL rows in a VAC crew group (sorted, deduplicated, comma-joined) instead of reading only from `sorted_rows[0]`. This ensures any per-row VAC crew field edit changes the hash and triggers regeneration.

- **Cache invalidation**: Bumped `DISCOVERY_CACHE_VERSION` from 2 → 3 to force re-validation of existing discovery caches. This ensures sheets that had VAC crew columns added after v2 will properly pick up those mappings on the next run rather than waiting up to 7 days for automatic refresh.

- **Regression tests**: Added `TestVacCrewHashAggregation` test class with three test cases:
  - `test_hash_changes_when_non_first_row_vac_crew_dept_edited`: Verifies hash changes when dept is edited on a non-first row
  - `test_hash_changes_when_non_first_row_vac_crew_name_edited`: Verifies hash changes when name is edited on a non-first row
  - `test_hash_stable_when_no_vac_crew_fields_change`: Verifies hash is stable and order-independent when no VAC crew fields change

## Implementation Details

The fix recognizes that VAC crew groups use a single key (`{week}_{wr}_VACCREW`) without per-foreman disambiguation, unlike helper groups which use `{week}_{wr}_HELPER_{helper}`. This means a single group contains all VAC crew members for a WR+week combination. The aggregation approach (sorted sets, deduplicated, comma-joined) maintains hash stability while capturing all relevant metadata changes.

Documentation of the bug, fix, and new rules has been added to CLAUDE.md for future reference.

https://claude.ai/code/session_018XPtK1XraGyJHTfBUoUKB2

---

> [!NOTE]
> **Medium Risk**
> Touches production-critical change-detection hashing and cache invalidation, which can force or skip Excel regeneration if wrong; added regression tests reduce the likelihood of unintended churn or missed updates.
> 
> **Overview**
> Fixes VAC crew Excel regeneration being incorrectly skipped when editing VAC crew name/dept/job on a non-first row by updating `calculate_data_hash()` to include VAC crew fields **per row** (only for the `vac_crew` variant) and adding VAC-crew tie-breakers to the extended-mode sort for deterministic hashing.
> 
> Bumps `DISCOVERY_CACHE_VERSION` to invalidate stale sheet/column mapping caches, and adds focused regression coverage in `tests/test_vac_crew.py` for non-first-row edits, set-dedup and comma-in-name collision edge cases, legacy-mode stability, and insertion-order stability. Documentation in `CLAUDE.md` records the incident, fix, and new rules for future hashing changes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit fe2730408bc8bdf58994aecdb8a6852654f3e47d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by JFlo21** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/jflo21/pr-checks/44198339-2830-48f4-83e3-f54278537e0f?refs=fe2730408bc8bdf58994aecdb8a6852654f3e47d&source=prChecksComment) |
| :white_check_mark: | Code Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/jflo21/project/eb083177-20b4-46e7-9447-b20c1e497e41/pr-check/908b2fa3-1b92-4bd2-9a8a-0d855f438dad?refs=fe2730408bc8bdf58994aecdb8a6852654f3e47d&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: generate_weekly_pdfs.py [hunk 1]
```diff
@@ -1227,15 +1230,26 @@ def calculate_data_hash(group_rows: list[dict]) -> str:
...
+        })
+        vac_crew_depts = sorted({
+            str(r.get('__vac_crew_dept') or '') for r in sorted_rows
+        })
+        vac_crew_jobs = sorted({
+            str(r.get('__vac_crew_job') or '') for r in sorted_rows
+        })
+        meta_parts.append(f"VACCREW={','.join(vac_crew_names)}")
+        meta_parts.append(f"VACCREW_DEPT={','.join(vac_crew_depts)}")
+        meta_parts.append(f"VACCREW_JOB={','.join(vac_crew_jobs)}")
```
[the remark this answers is not in the document]
Comment (JFlo21) [reply generate_weekly_pdfs.py]: Good catch. Fixed in e93d427 by moving VAC crew fields into the per-row `row_str` hash instead of a joined string — each row's name/dept/job is hashed independently, so `['A,B','C']` and `['A','B,C']` can no longer collapse to the same token regardless of what delimiter is chosen. Regression covered by `test_hash_distinguishes_comma_in_name_edge_case`.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (JFlo21) [reply generate_weekly_pdfs.py]: Confirmed and fixed in e93d427. You're right that the `meta_parts` set was the only place VAC crew fields reached the hash, so depts `{500, 500, 600}` + an edit from 500→600 silently skipped regeneration.

The fix moves the three VAC crew fields into the per-row `row_str` (scoped to the `vac_crew` variant so primary/helper hash stability is preserved) instead of aggregating them. Each row contributes its own name/dept/job independently, so the `{500, 500, 600}` → `{500, 600, 600}` edit now produces a distinct `row_str` for that row and a different group hash. The `meta_parts` aggregation is removed since per-row inclusion subsumes it.

Regression: `test_hash_changes_when_edited_value_collides_with_another_member` uses the exact three-member dept-collision scenario you described.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: CLAUDE.md [hunk 2]
```diff
@@ -396,3 +396,42 @@ When proposing new workflows, dynamically evaluate the absolute best technology.
   and the safety guard that still retains SmartSheet price when
   neither group nor CU is in new rates
   (`test_silent_fallthrough_when_neither_group_nor_cu_in_new_rates`).
+- [2026-04-22 00:00] VAC crew Excel files silently not regenerating
+  when a non-first-sorted row's VAC crew fields change. **Root
+  cause:** `calculate_data_hash()` builds `vac_crew` variant
+  metadata (VACCREW / VACCREW_DEPT / VACCREW_JOB) from
+  `sorted_rows[0]` only — mirroring the helper pattern — but the
+  `vac_crew` group key at `generate_weekly_pdfs.py:2660`
```
[the remark this answers is not in the document]
Comment (JFlo21) [reply CLAUDE.md]: Agreed — fixed in e93d427. The ledger entry no longer hard-codes line numbers anywhere; it now references the `_VACCREW` group-key format and function names (e.g. `group_source_rows`) instead. Also added a general rule to the entry that future living-ledger entries and code comments in this repo must refer to functions / group-key formats / env vars rather than line numbers.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/test_vac_crew.py [hunk 3]
```diff
@@ -224,5 +224,79 @@ def test_vac_crew_group_created_log_fires_once_per_group(self):
...
+            'CU': cu,
+            'Quantity': qty,
+            'Units Total Price': price,
+            'Units Completed?': True,
+            '__variant': 'vac_crew',
+            '__is_vac_crew': True,
+            '__vac_crew_name': name,
+            '__vac_crew_dept': dept,
+            '__vac_crew_job': job,
+        }
```
[the remark this answers is not in the document]
Comment (JFlo21) [reply tests/test_vac_crew.py:274]: Done in e93d427. `TestVacCrewHashAggregation` now has `setUp` / `tearDown` that save and restore `EXTENDED_CHANGE_DETECTION`, `RATE_CUTOFF_DATE`, and `_RATES_FINGERPRINT`, and force them to the production-default values (`True`, `None`, `''`) for the duration of each test. Developer env-var overrides can no longer destabilize the suite.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: generate_weekly_pdfs.py [hunk 4]
```diff
@@ -1227,15 +1230,26 @@ def calculate_data_hash(group_rows: list[dict]) -> str:
...
-        meta_parts.append(f"VACCREW_DEPT={vac_crew_dept}")
-        meta_parts.append(f"VACCREW_JOB={vac_crew_job}")
+        # VAC Crew variant is NOT split per foreman in the group key
+        # (line 2660 uses `_VACCREW` with no foreman suffix), so a single
+        # group can contain multiple VAC crew members. Reading only
+        # sorted_rows[0] would miss hash changes when VAC crew
+        # fields (name/dept/job) are edited on any row that is not first
+        # in sort order — the known "VAC crew sheet not regenerating even
+        # though criteria is met" bug. Aggregate across all rows so any
+        # per-row VAC crew field change forces a new hash.
```
[the remark this answers is not in the document]
Comment (JFlo21) [reply generate_weekly_pdfs.py]: Addressed in e93d427. The block you flagged was removed entirely — VAC crew fields now flow through the per-row `row_str` instead of a separate `meta_parts` aggregation, which also fixes the set-dedup and delimiter-collision bugs the other reviewers caught. The replacement inline comment (next to the per-row extension) describes the `_VACCREW` group-key shape and the two pitfalls by name, with no line numbers.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: generate_weekly_pdfs.py [hunk 5]
```diff
@@ -1195,7 +1203,27 @@ def calculate_data_hash(group_rows: list[dict]) -> str:
...
+        #      regeneration.
+        #   2. Delimiter collision — ','.join on free-text names cannot
+        #      distinguish ['A,B', 'C'] from ['A', 'B,C'].
+        # Scoped to vac_crew so hash stability for primary/helper rows is
+        # preserved (non-vac_crew row_str structure is unchanged).
+        if group_variant == 'vac_crew':
+            row_fields.extend([
+                str(row.get('__vac_crew_name') or ''),
+                str(row.get('__vac_crew_dept') or ''),
+                str(row.get('__vac_crew_job') or ''),
```
[the remark this answers is not in the document]
Comment (JFlo21) [reply generate_weekly_pdfs.py:1247]: Good catch. Fixed in 52643f0 by adding `__vac_crew_name`, `__vac_crew_dept`, and `__vac_crew_job` as tie-breakers in the existing `sorted(...)` key inside `calculate_data_hash()`. For non-VAC-crew rows those fields are empty strings, so the tie-breaker resolves to `('', '', '')` — a no-op that preserves hash stability for all primary/helper groups. For VAC-crew rows tied on `(WR, Snapshot Date, CU, Pole, Qty)` the tie-breaker now disambiguates deterministically, so the `as_completed`-induced insertion-order variance can no longer bleed into the hash.

Regression: `test_hash_stable_when_vac_crew_rows_share_primary_sort_keys` uses two crew members with identical primary-sort-key tuples to exercise the tie-breaker directly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: generate_weekly_pdfs.py [hunk 6]
```diff
@@ -1144,6 +1155,9 @@ def calculate_data_hash(group_rows: list[dict]) -> str:
             str(x.get('CU', '')),
             str(x.get('Pole #') or x.get('Point #') or x.get('Point Number') or ''),
             str(x.get('Quantity', '')),
+            str(x.get('__vac_crew_name') or ''),
+            str(x.get('__vac_crew_dept') or ''),
+            str(x.get('__vac_crew_job') or ''),
```
[the remark this answers is not in the document]
Comment (JFlo21) [reply generate_weekly_pdfs.py]: Good point — fixed in fe27304. The tie-breaker is now gated behind `EXTENDED_CHANGE_DETECTION`: legacy mode keeps its original 5-key sort so hashes stay bit-identical to pre-fix and the docstring's rollback-stability guarantee is preserved; extended mode (production default) still gets the VAC crew tie-breaker to neutralize `as_completed`-induced insertion-order variance. New regression `test_legacy_mode_ignores_vac_crew_metadata_edits` confirms a VAC crew field edit on rows tied on primary sort keys (with differing `Work Type`) leaves the legacy hash unchanged.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (JFlo21) [reply generate_weekly_pdfs.py]: Agreed — fixed in fe27304. The tie-breakers are now gated behind `EXTENDED_CHANGE_DETECTION`, so legacy mode (`EXTENDED_CHANGE_DETECTION=0`) keeps its original 5-key sort and the minimal-fields / rollback-stability guarantee stays intact. VAC crew metadata edits can no longer reorder legacy rows or perturb legacy hashes indirectly. Regression: `test_legacy_mode_ignores_vac_crew_metadata_edits` pins `EXTENDED_CHANGE_DETECTION=False`, uses two rows tied on primary sort keys with differing `Work Type`, edits a VAC crew field on one of them, and asserts the legacy hash is unchanged.

---
_Generated by [Claude Code](https://claude.ai/code)_
